shell/platform/android/io/flutter/view/FlutterNativeView.java

```diff
@@ -149,77 +161,89 @@ public void setMessageHandler(String channel, BinaryMessageHandler handler) {
         }
-        nativeInvokePlatformMessageEmptyResponseCallback(mNativePlatformView, replyId);
-    }
 
-    // Called by native to respond to a platform message that we sent.
-    private void handlePlatformMessageResponse(int replyId, byte[] reply) {
-        BinaryReply callback = mPendingReplies.remove(replyId);
-        if (callback != null) {
-            try {
-                callback.reply(reply == null ? null : ByteBuffer.wrap(reply));
-            } catch (Exception ex) {
-                Log.e(TAG, "Uncaught exception in binary message reply handler", ex);
+        // Called by native to respond to a platform message that we sent.
+        public void handlePlatformMessageResponse(int replyId, byte[] reply) {
+            BinaryReply callback = mPendingReplies.remove(replyId);
+            if (callback != null) {
+                try {
+                    callback.reply(reply == null ? null : ByteBuffer.wrap(reply));
+                } catch (Exception ex) {
+                    Log.e(TAG, "Uncaught exception in binary message reply handler", ex);
+                }
             }
         }
     }
 
-    // Called by native to update the semantics/accessibility tree.
-    private void updateSemantics(ByteBuffer buffer, String[] strings) {
-        if (mFlutterView == null) return;
-        mFlutterView.updateSemantics(buffer, strings);
-    }
+    private final class RenderSurfaceImpl implements RenderSurface {
+        // Called by native to update the semantics/accessibility tree.
+        public void updateSemantics(ByteBuffer buffer, String[] strings) {
+            if (mFlutterView == null) return;
+            mFlutterView.updateSemantics(buffer, strings);
+        }
 
-    // Called by native to update the custom accessibility actions.
-    private void updateCustomAccessibilityActions(ByteBuffer buffer, String[] strings) {
-        if (mFlutterView == null)
-            return;
-        mFlutterView.updateCustomAccessibilityActions(buffer, strings);
-    }
+        // Called by native to update the custom accessibility actions.
+        public void updateCustomAccessibilityActions(ByteBuffer buffer, String[] strings) {
+            if (mFlutterView == null)
+                return;
+            mFlutterView.updateCustomAccessibilityActions(buffer, strings);
+        }
+
+        // Called by native to notify first Flutter frame rendered.
+        public void onFirstFrameRendered() {
+            if (mFlutterView == null) return;
+            mFlutterView.onFirstFrame();
+        }
 
-    // Called by native to notify first Flutter frame rendered.
-    private void onFirstFrame() {
-        if (mFlutterView == null) return;
-        mFlutterView.onFirstFrame();
+        /*package*/ FlutterJNI getFlutterJNI() {
+            return mFlutterJNI;
+        }
     }
 
     // Called by native to notify when the engine is restarted (cold reload).
```

Comment:
This also needs to be updated, just noticed it. It's being wired in the JNI layer to `FlutterEngine.java` still. It can be updated to work through an interface in `FlutterJNI` the same way `PlatformMessageHandler` and `FlutterRenderer.RenderSurface` do, I'll fix that in a new commit. Good news is that it's very easily testable, called on hot restart.

Reply:
Would also mind rebasing off of my PR branch?  That branch should be merged in tomorrow and it's possible it introduces conflict with some of this stuff.

Reply:
Fixed.

I think the diff in this PR would be really hard to read if I rebased it right now, since this is being compared to and merged into `flutter/engine:21008_rewrite-android-embedding-to-reduce-coupling`. I can wait on submitting this until after flutter/engine#6399 and update this then if you'd rather though.

Reply:
Yeah I think it would be helpful if you waited.  I think it'll be easier for you to adjust JNI to match the changes to 21008 than it would be in the reverse.